File: python/services/ai/mol/providers/openai.py
"""OpenAI provider — async HTTP client for the Chat Completions API.

Mirrors :file:`supabase/functions/_shared/mol/providers/openai.ts`:
- POST to https://api.openai.com/v1/chat/completions
- ``Authorization: Bearer <key>`` header
- System prompt as the first ``role: system`` message
- Vision: ``image_url`` content part on the last user message
- Tokens read from ``usage.prompt_tokens`` / ``usage.completion_tokens``
- Finish reason from ``choices[0].finish_reason``
"""

from __future__ import annotations

import time
from typing import Any

import httpx

from ...config import get_settings
from ..types import ChatTurn, ProviderResponse, TokenUsage
from .base import ModelProvider

OPENAI_URL = "https://api.openai.com/v1/chat/completions"


class OpenAIError(Exception):
    """Wraps OpenAI API failures with a status-encoded message string."""

    def __init__(self, status: int, typed_label: str = "") -> None:
        suffix = f" ({typed_label})" if typed_label else ""
        super().__init__(f"OpenAI {status}{suffix}")
        self.status = status


class OpenAIProvider(ModelProvider):
    id = "openai"
    default_model = "gpt-4o-mini"

    def _api_key(self) -> str:
        return get_settings().openai_api_key

    def is_configured(self) -> bool:
        return bool(self._api_key())

    async def call(
        self,
        *,
        model: str,
        system_prompt: str,
        user_messages: list[ChatTurn],
        max_tokens: int,
        temperature: float = 0.7,
        timeout_seconds: int = 20,
        image_url: str | None = None,
    ) -> ProviderResponse:
        if not self.is_configured():
            raise RuntimeError("OpenAIProvider not configured (OPENAI_API_KEY missing)")

        chat_messages: list[dict[str, Any]] = [
            {"role": "system", "content": system_prompt},
        ]
        for i, m in enumerate(user_messages):
            is_last = i == len(user_messages) - 1
            if is_last and image_url:
                chat_messages.append(
                    {
                        "role": m.role,
                        "content": [
                            {"type": "image_url", "image_url": {"url": image_url}},
                            {"type": "text", "text": m.content},
                        ],
                    }
                )
            else:
                chat_messages.append({"role": m.role, "content": m.content})

        body = {
            "model": model,
            "messages": chat_messages,
            "max_tokens": max_tokens,
            "temperature": temperature,
        }

        headers = {
            "Authorization": f"Bearer {self._api_key()}",
            "Content-Type": "application/json",
        }

        start = time.monotonic()
        async with httpx.AsyncClient(timeout=timeout_seconds) as client:
            res = await client.post(OPENAI_URL, json=body, headers=headers)

        if res.status_code >= 400:
            typed = ""
            try:
                j = res.json()
                err = j.get("error") if isinstance(j, dict) else None
                if isinstance(err, dict):
                    tag = err.get("code") or err.get("type")
                    if tag:
                        typed = str(tag)
            except Exception:  # noqa: BLE001 — defensive parity with TS
                pass
            raise OpenAIError(res.status_code, typed)

        data = res.json()
        choices = data.get("choices") or []
        first = choices[0] if choices else {}
        message = (first.get("message") or {}) if isinstance(first, dict) else {}
        text = (message.get("content") or "").strip()
        usage = data.get("usage") or {}

        latency_ms = int((time.monotonic() - start) * 1000)
        return ProviderResponse(
            text=text,
            provider="openai",
            model=str(data.get("model") or model),
            tokens=TokenUsage(
                prompt=int(usage.get("prompt_tokens", 0)),
                completion=int(usage.get("completion_tokens", 0)),
            ),
            finish_reason=str(first.get("finish_reason") or "stop"),
            raw={"latency_ms": latency_ms, **(data if isinstance(data, dict) else {})},
        )

File: supabase/functions/_shared/mol/providers/openai.ts
// supabase/functions/_shared/mol/providers/openai.ts

import type { ModelProvider, ProviderCallOptions } from './base.ts'
import type { ProviderResponse } from '../types.ts'
import { fetchWithTimeout } from '../../reliability.ts'

const OPENAI_URL = 'https://api.openai.com/v1/chat/completions'

export class OpenAIProvider implements ModelProvider {
  readonly id = 'openai' as const
  readonly default_model = 'gpt-4o-mini'

  private apiKey(): string {
    return Deno.env.get('OPENAI_API_KEY') || ''
  }

  isConfigured(): boolean {
    return this.apiKey().length > 0
  }

  async call(model: string, opts: ProviderCallOptions): Promise<ProviderResponse> {
    if (!this.isConfigured()) {
      throw new Error('OpenAIProvider not configured (OPENAI_API_KEY missing)')
    }

    const timeout = opts.timeout_ms ?? 20_000

    // OpenAI chat format. Vision via image_url content part on the last user msg.
    const chatMessages: Array<Record<string, unknown>> = [
      { role: 'system', content: opts.system_prompt },
    ]
    for (let i = 0; i < opts.user_messages.length; i++) {
      const m = opts.user_messages[i]
      const isLast = i === opts.user_messages.length - 1
      if (isLast && opts.image_url) {
        chatMessages.push({
          role: m.role,
          content: [
            { type: 'image_url', image_url: { url: opts.image_url } },
            { type: 'text', text: m.content },
          ],
        })
      } else {
        chatMessages.push({ role: m.role, content: m.content })
      }
    }

    const body = {
      model,
      messages: chatMessages,
      max_tokens: opts.max_tokens,
      temperature: opts.temperature ?? 0.7,
    }

    const res = await fetchWithTimeout(OPENAI_URL, {
          provider: 'openai',
          operation: 'mol_chat_completion',
          timeoutMs: timeout,
          method: 'POST',
          headers: {
            'Authorization': `Bearer ${this.apiKey()}`,
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(body),
        })

    if (!res.ok) {
      const raw = await res.text().catch(() => '')
      let typed = ''
      try {
        const j = JSON.parse(raw) as { error?: { type?: string; code?: string } }
        const tag = j?.error?.code || j?.error?.type
        if (tag) typed = ` (${tag})`
      } catch { /* not JSON */ }
      throw new Error(`OpenAI ${res.status}${typed}`)
    }

    const data = await res.json() as {
      choices: Array<{ message: { content: string }; finish_reason: string }>
      usage: { prompt_tokens: number; completion_tokens: number }
      model: string
    }

    const text = (data.choices[0]?.message?.content ?? '').trim()

    return {
      text,
      provider: 'openai',
      model: data.model || model,
      tokens: { prompt: data.usage.prompt_tokens, completion: data.usage.completion_tokens },
      finish_reason: data.choices[0]?.finish_reason ?? 'stop',
      raw: data,
    }
  }
}
